File: app/src/main/java/de/hype/hypenotify/services/HypeNotifyService.java
package de.hype.hypenotify.services;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import androidx.annotation.Nullable;
import de.hype.hypenotify.core.interfaces.MiniCore;

public class HypeNotifyService<EXTENDING_CLASS extends HypeNotifyService<EXTENDING_CLASS>> extends Service {
    protected MiniCore core;
    protected Context context;

    @Nullable
    @Override
    @SuppressWarnings("unchecked")
    public IBinder onBind(Intent intent) {
        return new HypeNotifyServiceBinder((EXTENDING_CLASS) this);
    }


    public MiniCore getCore() {
        return core;
    }

    public class HypeNotifyServiceBinder extends Binder {
        private final EXTENDING_CLASS service;

        public HypeNotifyServiceBinder(EXTENDING_CLASS service) {
            this.service = service;
        }

        public EXTENDING_CLASS getService() {
            return service;
        }
    }
}

File: app/src/main/java/de/hype/hypenotify/app/services/HypeNotifyService.kt
package de.hype.hypenotify.app.services

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import de.hype.hypenotify.app.core.interfaces.MiniCore

open class HypeNotifyService<EXTENDING_CLASS : HypeNotifyService<EXTENDING_CLASS>> : Service() {
    open var core: MiniCore? = null
        protected set
    protected var context: Context? = null

    override fun onBind(intent: Intent): IBinder {
        return HypeNotifyServiceBinder(this as EXTENDING_CLASS)
    }


    inner class HypeNotifyServiceBinder(val service: EXTENDING_CLASS) : Binder() {
    }
}
